### PR TITLE
Ruby 2.4 compatibility

### LINKED ISSUE #13: error: ‘rb_cFixnum’ undeclared
Linux x64 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:10:15 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-linux]

$ cd /usr/local/src/ruby-gmp-master/ext

$ /usr/local/src/ruby/bin/ruby extconf.rb
checking for gmp.h... yes
checking for __gmpz_init() in -lgmp... yes
checking for mpfr.h... yes
checking for mpf2mpfr.h... yes
checking for mpfr_init() in -lmpfr... yes
checking for SIZEOF_INTPTR_T... yes
creating Makefile

$ make
compiling gmp.c
gmp.c: In function ‘Init_gmp’:
gmp.c:165:19: error: ‘rb_cFixnum’ undeclared (first use in this function)
   rb_define_alias(rb_cFixnum, old_fname, ruby_fname); \
                   ^
gmp.c:222:3: note: in expansion of macro ‘REGISTER_TAKEOVER’
   REGISTER_TAKEOVER(and, "&", "old_and")
   ^
gmp.c:165:19: note: each undeclared identifier is reported only once for each function it appears in
   rb_define_alias(rb_cFixnum, old_fname, ruby_fname); \
                   ^
gmp.c:222:3: note: in expansion of macro ‘REGISTER_TAKEOVER’
   REGISTER_TAKEOVER(and, "&", "old_and")
   ^
gmp.c:167:19: error: ‘rb_cBignum’ undeclared (first use in this function)
   rb_define_alias(rb_cBignum, old_fname, ruby_fname); \
                   ^
gmp.c:222:3: note: in expansion of macro ‘REGISTER_TAKEOVER’
   REGISTER_TAKEOVER(and, "&", "old_and")
   ^
gmp.c: At top level:
cc1: warning: unrecognized command line option ‘-Wno-self-assign’
cc1: warning: unrecognized command line option ‘-Wno-constant-logical-operand’
cc1: warning: unrecognized command line option ‘-Wno-parentheses-equality’
cc1: warning: unrecognized command line option ‘-Wno-tautological-compare’
Makefile:241: recipe for target 'gmp.o' failed
make: *** [gmp.o] Error 1

### PR DESCRIPTION
**Pull request by timocp**
- Added support for integer unification in ruby 2.4 (Fixnum/Bignum become Integer)
- Test::Unit changed in ruby 2.2 so switched to MiniTest
- In ruby 2.4, can't tell size of fixnums, now always run 32/64 bit tests
- 2nd param to `Z#addmul!` and `Z#submul!` can now be small negatives

Closes #13

### CODE REVIEW

**Review by srawlins — COMMENTED** [review]
Thanks for this PR! This is a ton of work; you don't know how much I appreciate this 😄!  Reviewing now...

File: ext/gmpz.c [hunk 1]
```diff
@@ -1177,14 +1177,9 @@ static VALUE r_gmpz_addmul_self(VALUE self, VALUE b, VALUE c)
...
-  } else if (TYPE (c) == T_FIXNUM) {
-    if (FIX2NUM (c) < 0)
-    {
-      if (free_b_val) { mpz_temp_free (b_val); }
-      rb_raise (rb_eRangeError, "multiplicand (Fixnum) must be nonnegative");
-    }
+  } else if (FIXNUM_P (c) && FIX2NUM (c) >= 0) {
     mpz_addmul_ui (self_val, b_val, FIX2NUM (c));
-  } else if (BIGNUM_P (c)) {
+  } else if (FIXNUM_P (c) || BIGNUM_P (c)) {
```
Comment (srawlins) [inline ext/gmpz.c]: Nice fix.

File: test/gmp_tcong.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 require File.expand_path(File.join(File.dirname(__FILE__), 'test_helper'))
 
-class GMP_TCong < Test::Unit::TestCase
+class GMP_TCong < MiniTest::Test
```
Comment (srawlins) [inline test/gmp_tcong.rb]: This is also a big win. Thanks.

**Review by srawlins — APPROVED** [review]
Looks great! Thanks again!

File: test/gmp_tcong.rb (hunk 2 above)
Comment (timocp) [reply test/gmp_tcong.rb:3]: Cool - this seemed to be the easiest way to ensure the tests run in all versions of ruby from 1.9 to 2.4 :)

**Pull request merged**